SimG4CMS/Forward/plugins/DoCastorAnalysis.cc

```diff
@@ -9,24 +9,104 @@
 // Original Author: P. Katsas
 //         Created: 02/2007
 //
+
+#include "DataFormats/Math/interface/Point3D.h"
+
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
 #include "SimG4CMS/Calo/interface/CaloG4Hit.h"
 #include "SimG4CMS/Calo/interface/CaloG4HitCollection.h"
-#include "DataFormats/Math/interface/Point3D.h"
-
-#include "SimG4CMS/Forward/interface/DoCastorAnalysis.h"
 #include "SimG4CMS/Forward/interface/CastorNumberingScheme.h"
 
+#include "SimG4Core/Notification/interface/BeginOfJob.h"
+#include "SimG4Core/Notification/interface/BeginOfRun.h"
+#include "SimG4Core/Notification/interface/EndOfRun.h"
+#include "SimG4Core/Notification/interface/BeginOfEvent.h"
+#include "SimG4Core/Notification/interface/EndOfEvent.h"
+#include "SimG4Core/Notification/interface/Observer.h"
+#include "SimG4Core/Watcher/interface/SimWatcher.h"
+
+#include "G4SDManager.hh"
+#include "G4Step.hh"
+#include "G4Track.hh"
+#include "G4Event.hh"
+#include "G4PrimaryVertex.hh"
+#include "G4VProcess.hh"
+#include "G4HCofThisEvent.hh"
+#include "G4UserEventAction.hh"
+#include "CLHEP/Units/GlobalSystemOfUnits.h"
+#include "CLHEP/Units/GlobalPhysicalConstants.h"
+#include "CLHEP/Random/Randomize.h"
+
+#include "TROOT.h"
 #include "TFile.h"
+#include "TH1.h"
+#include "TH2.h"
+#include "TProfile.h"
+#include "TNtuple.h"
+#include "TRandom.h"
+#include "TLorentzVector.h"
+#include "TUnixSystem.h"
+#include "TSystem.h"
+#include "TMath.h"
+#include "TF1.h"
+
+#include <cassert>
 #include <cmath>
 #include <iostream>
 #include <iomanip>
+#include <map>
 #include <memory>
+#include <string>
 #include <vector>
 
 //#define EDM_ML_DEBUG
 
-DoCastorAnalysis::DoCastorAnalysis(const edm::ParameterSet& p) {
+class DoCastorAnalysis : public SimWatcher,
+                         public Observer<const BeginOfJob *>,
+                         public Observer<const BeginOfRun *>,
+                         public Observer<const EndOfRun *>,
+                         public Observer<const BeginOfEvent *>,
+                         public Observer<const EndOfEvent *>,
+                         public Observer<const G4Step *> {
+public:
+  DoCastorAnalysis(const edm::ParameterSet &p);
+  ~DoCastorAnalysis() override;
+
+private:
+  // observer classes
+  void update(const BeginOfJob *run) override;
+  void update(const BeginOfRun *run) override;
+  void update(const EndOfRun *run) override;
+  void update(const BeginOfEvent *evt) override;
+  void update(const EndOfEvent *evt) override;
+  void update(const G4Step *step) override;
+
+private:
+  int verbosity;
+
+  std::string TreeFileName;
+
+  TFile *CastorOutputEventFile;
+  TTree *CastorTree;
+
+  int eventIndex;
+
+  std::vector<double> simhit_x, simhit_y, simhit_z;
+  std::vector<double> simhit_eta, simhit_phi, simhit_energy;
+  std::vector<int> simhit_sector, simhit_module;
+  //std::vector<double> simhit_time;
```

Comment:
Should `simhit_time` and `*psimhit_time` remain here (for possible future usages)?
If not, better to remove these commented out definitions, and the corresponding commented out lines in the implementation part.

Reply:
Taken care of